packages/mesos-client/src/rxjs-extensions/adaptiveTimeout.js

```diff
@@ -0,0 +1,89 @@
+import { Subscriber, Scheduler, TimeoutError } from "rxjs";
+
+// As Mesos documentation suggests, to make sure the connection is dead
+// we should wait for "an extended period of time" e.g., 5 x heartbeat_interval_seconds
+export const TIMEOUT_FACTOR = 5;
+const DEFAULT_TIMEOUT = 5000;
+
+export default function adaptiveTimeout(
```

Comment:
Do you think there is a way to write this operator so that it's generally applicable?

Reply:
Sure, by exposing the update function as an argument. Do we really need this?
So the function can be called on every `next` passing the value inside. Whether use this value or not it's up to the function.

Reply:
rxjs actually had a `timeoutWithSelector` which was dropped in version 5 to reduce weight.

http://reactivex.io/documentation/operators/timeout.html

I'm fine if we don't generalize the operator as a simple `timeout` is probably sufficient in most cases, but then let's adjust the name so that it's obvious that this one checks the mesos heartbeat.

Reply:
After some meditation on the class based operators I think it's not a good way to go. It exposes a lot of rxjs internals that 1) can change over time 2) one needs to know to understand what's happening.

how about something like this:
```javascript
function mesosTimeout(source) {
  let heartBeat;
  
  const sourceSubject = source.publish().refCount();
  
  return Rx.Observable.create(function subscribe(observer) {
    // First observable reads timeout and completes
    const sub = sourceSubject.do(function(value) {
      heartBeat = 1100;
      observer.next(value);
      observer.complete();
    }).subscribe();
    
    return function() {
      sub.unsubscribe();
    }
  }).concat(
    Rx.Observable.create(function (observer) {      
      // Second observable uses previously read timeout to control the stream
      sourceSubject
        .timeout(heartBeat)
        .subscribe(function(value) {
          observer.next(value);
        });
    })
  );
} 

mesosTimeout(Rx.Observable.interval(1000))
  .subscribe(function(value){ 
    console.log(value);
  });
```

Reply:
@MatApple would like to get your view on both versions.

Reply:
I'd like to prose yet another solution to monitor the  "heartbeat" without any side effects. It also features a selector to dynamically adapt the timeout.

```js
// monitor heartbeat
function monitor(source, timeoutSelector = function(){}, initialTimeout=500){
   return source
     .scan(
        ({timeout}, message) => {
        const _timeout = timeoutSelector(message);
          
        if(!Number.isFinite(_timeout)){
          return {message, timeout: _timeout}
         }
          
        return {message, timeout};     
       }, {timeout: initialTimeout})
     .map(({timeout, message}) => { 
       return Rx.Observable.create(function create(observer){
         const timeoutId = setTimeout(()=> observer.error(), timeout);

         // pass message along 
         observer.next(message);

         return function disspose(){
           clearTimeout(timeoutId); 
         };                       
       })
     })
    .switch()
}
```

**Example**
```js
// Debug Source
const source = Rx.Observable.of(1000,500,1000,800,500,1000, 2000, 5000, 1000, 6000)
   .map((time)=>Rx.Observable.timer(time))
   .concatAll()
   .map((_, index ) => { 
     if(index === 2){
       return {index, heartbeat: 100}; 
     }
     return {index}; 
   });

    monitor(source, (message) => message.heatbeat * 5, 6000
     .subscribe(
          (message)=>console.log("message", message),
          (error)=>console.log("error", error)
      )
```

Reply:
The implementation is elegant though it unfortunately doesn't work with an initial timeout, and I'd rather avoid `setTimeout` in favor of native timeouts that work on the rxjs scheduler thus can be tested with built-in tooling.